File: zeppelin-examples/zeppelin-example-horizontalbar/src/main/java/org/apache/zeppelin/example/app/horizontalbar/HorizontalBar.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.zeppelin.example.app.horizontalbar;

import org.apache.commons.io.IOUtils;
import org.apache.zeppelin.helium.Application;
import org.apache.zeppelin.helium.ApplicationContext;
import org.apache.zeppelin.helium.ApplicationException;
import org.apache.zeppelin.interpreter.InterpreterResult;
import org.apache.zeppelin.interpreter.dev.ZeppelinApplicationDevServer;
import org.apache.zeppelin.resource.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.io.InputStream;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Random;

/**
 * Basic example application.
 * TableData for input
 */
public class HorizontalBar extends Application {
  private final Logger logger = LoggerFactory.getLogger(HorizontalBar.class);

  InterpreterResult result;

  public HorizontalBar(ApplicationContext context) {
    super(context);
  }

  @Override
  public void run(ResourceSet resources) throws ApplicationException, IOException {
    // Get data from resource args
    result = (InterpreterResult) resources.get(0).get();

    // create element
    println(String.format(
        "<div id=\"horizontalbar_%s\" style=\"height:400px\"><svg></svg></div>",
        context().getApplicationInstanceId()));
    // write js
    printResourceAsJavascript("example/app/horizontalbar/horizontalbar.js");
  }

  @Override
  public void unload() throws ApplicationException {
  }

  /**
   * Development mode
   */
  public static void main(String[] args) throws Exception {
    LocalResourcePool pool = new LocalResourcePool("dev");
    InputStream ins = ClassLoader.getSystemResourceAsStream(
        "example/app/horizontalbar/horizontalbar_mockdata.txt");
    InterpreterResult result = new InterpreterResult(
        InterpreterResult.Code.SUCCESS,
        InterpreterResult.Type.TABLE,
        IOUtils.toString(ins));
    pool.put(WellKnownResourceName.ZeppelinTableResult.name(), result);

    ZeppelinApplicationDevServer devServer = new ZeppelinApplicationDevServer(
        HorizontalBar.class.getName(),
        pool.getAll());

    devServer.start();
    devServer.join();
  }
}

File: zeppelin-examples/zeppelin-example-horizontalbar/horizontalbar.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import Nvd3ChartVisualization from 'zeppelin-vis/builtins/visualization-nvd3chart';
import PivotTransformation from 'zeppelin-tabledata/pivot';

/**
 * Base class for visualization
 */
export default class horizontalbar extends Nvd3ChartVisualization {
  constructor(targetEl, config) {
    super(targetEl, config)
    this.pivot = new PivotTransformation(config);
  }

  type() {
    return 'multiBarHorizontalChart';
  };

  render(pivot) {
    var d3Data = this.d3DataFromPivot(
      pivot.schema,
      pivot.rows,
      pivot.keys,
      pivot.groups,
      pivot.values,
      true,
      false,
      true);

    super.render(d3Data);
  }

  getTransformation() {
    return this.pivot;
  }
  
  /**
   * Set new config
   */
  setConfig(config) {
    super.setConfig(config);
    this.pivot.setConfig(config);
  };

  configureChart(chart) {
    var self = this;
    chart.yAxis.axisLabelDistance(50);
    chart.yAxis.tickFormat(function(d) {return self.yAxisTickFormat(d);});

    this.chart.stacked(this.config.stacked);

    var self = this;
    this.chart.dispatch.on('stateChange', function(s) {
      self.config.stacked = s.stacked;

      // give some time to animation finish
      setTimeout(function() {
        self.emitConfig(self.config);
      }, 500);
    });
  };
}

